src/Microsoft.Diagnostics.ExtensionCommands/DumpStackObjectsCommand.cs

```diff
@@ -0,0 +1,381 @@
+            return null;
+        }
+
+        private IEnumerable<(ulong RegisterIndex, ulong PotentialObject)> EnumerateRegistersWithinHeapBounds()
+        {
+            ClrHeap heap = Runtime.Heap;
+
+            // Segments are always sorted by address
+            ulong minAddress = heap.Segments[0].ObjectRange.Start;
+            ulong maxAddress = heap.Segments[heap.Segments.Length - 1].ObjectRange.End - (uint)MemoryService.PointerSize;
+
+            int regCount = ThreadService.Registers.Count();
+            for (int i = 0; i < regCount; i++)
+            {
+                if (CurrentThread.TryGetRegisterValue(i, out ulong value))
+                {
+                    if (minAddress <= value && value < maxAddress)
+                    {
+                        yield return ((ulong)i, value);
+                    }
+                }
+            }
+        }
+
+        private IEnumerable<(ulong StackAddress, ulong PotentialObject)> EnumeratePointersWithinHeapBounds(MemoryRange stack)
+        {
+            Debug.Assert(AlignDown(stack.Start) == stack.Start);
+            Debug.Assert(AlignUp(stack.End) == stack.End);
+
+            uint pointerSize = (uint)MemoryService.PointerSize;
+            ClrHeap heap = Runtime.Heap;
+
+            // Segments are always sorted by address
+            ulong minAddress = heap.Segments[0].ObjectRange.Start;
+            ulong maxAddress = heap.Segments[heap.Segments.Length - 1].ObjectRange.End - pointerSize;
+
+            // Read in 64k chunks
+            byte[] buffer = ArrayPool<byte>.Shared.Rent(64 * 1024);
+            try
+            {
+                ulong address = stack.Start;
+                while (stack.Contains(address))
+                {
+                    Console.CancellationToken.ThrowIfCancellationRequested();
+
+                    if (!MemoryService.ReadMemory(address, buffer, out int read))
+                    {
+                        break;
+                    }
+
+                    read = AlignDown(read);
+                    if (read < pointerSize)
+                    {
+                        break;
+                    }
+
+                    for (int i = 0; i < read; i += (int)pointerSize)
+                    {
+                        Console.CancellationToken.ThrowIfCancellationRequested();
+
+                        ulong stackAddress = address + (uint)i;
+                        if (!stack.Contains(stackAddress))
+                        {
+                            yield break;
+                        }
+
+                        ulong potentialObj = GetIndex(buffer, i);
+                        if (minAddress <= potentialObj && potentialObj < maxAddress)
+                        {
+                            yield return (stackAddress, potentialObj);
+                        }
+                    }
+
+                    address += (uint)read;
+                }
+            }
+            finally
```

Comment:
Nit: the `finally` isn't strictly necessary here. In case of failure the rented array will just be dropped (and eligible for GC) which won't harm the ArrayPool in it's work.

In early days of ArrayPool the use of try-finally was broad, but in the runtime-repo they dropped the try-finally for this almost completely, as there's no need for it.

Reply:
It's not strictly required, no.  I would like to return the array if possible though, and the reason for the finally is that I use yield break.  I could duplicate the ArrayPool.Return call instead of having a finally though.  Either way.